src/components/library_card_table/LibraryCardTable.tsx

```diff
@@ -294,35 +295,51 @@ const LibraryCardTable: React.FC<LibraryCardTableProps> = (props) => {
   />;
 
   // onClick function, used to create new card on modal based on form data
-  const addLibraryCard = async (card: LibraryCard): Promise<void> => {
-    try {
-      // Check if card already exists, show error if it does
-      const alreadyExists = findIndex(
-          (element: LibraryCard) => isEqual(element.userEmail)(card.userEmail),
-          libraryCards
-      );
-      if (alreadyExists > -1) {
-        Notifications.showError({text: 'Library Card already exists'});
-      } else {
-        // Execute library card update with payload, get the updated card, and
-        // add (with sort afterwards) library card to libraryCards (reference list)
-        const newCard = await LibraryCardAPI.createLibraryCard(card);
-        const updatedList = cloneDeep(libraryCards);
-        updatedList.push(newCard);
-        updatedList.sort((a: LibraryCard, b: LibraryCard) => {
-          const dateA = new Date(a.createDate ?? '');
-          const dateB = new Date(b.createDate ?? '');
-          return dateB.getTime() - dateA.getTime();
-        });
-        setLibraryCards(updatedList);
-        setShowModal(false);
+  const addLibraryCards = async (newLibraryCards: LibraryCard[]): Promise<void> => {
+    // Check if any cards already exist, show error if any do
+    const duplicateLibraryCards = newLibraryCards.filter((card) => {
+        return findIndex(
+            (element: LibraryCard) => isEqual(element.userEmail)(card.userEmail),
+            libraryCards
+        ) > -1;
+    });
+
+    if (duplicateLibraryCards.length > 0) {
+      Notifications.showError({
+        text: 'Error updating library cards. The following users already have library cards: ' + duplicateLibraryCards.map(card => card.userEmail).join(', ')
+      });
+    } else {
+      // Execute library card updates with payload, get the updated card, and
+      // add (with sort afterwards) library card to libraryCards (reference list)
+
+      // Process each card individually and aggregate the results at the end.
+      // Eventually, if there's a bulk API endpoint, this can be optimized. But for
+      // now, the volume of cards being added is small enough that this should be okay
+      const successfulCards = [];
+      const failedCards = [];
+      for (const card of newLibraryCards) {
+        try {
+          const newCard = await LibraryCardAPI.createLibraryCard(card);
+          successfulCards.push(newCard);
+        } catch (error: unknown) {
+          const errorMessage = extractError(error);
+          failedCards.push({ card, error: errorMessage });
+        }
       }
```

Comment:
I could rewrite this to just fire these off all at once and then sort through the results once they're all complete, as long as there aren't any performance concerns with the underlying API?

Doing them serially doesn't seem slow at all from my testing, but happy to refactor this if anyone reviewing has a preference.

Reply:
Agree with Sky here. We do have something a bit old that I needed to ensure serial promise execution in [util.js](https://github.com/DataBiosphere/duos-ui/blob/develop/src/libs/utils.js#L200) but that might be more than what you want.